src/js/fable-core/Array2.fs

```diff
@@ -1,16 +1,732 @@
 module Array
 
+// This value is for use by compiled F# code and should not be used directly.
+// We can use LanguagePrimitives.ErrorStrings.
+// TODO: Approve please if we really need to use ErrorStrings here
+#nowarn "1204"
+
 open Fable.Core
+open Fable.Import
 
 type ArrayCons =
     [<Emit("new $0($1)")>]
     abstract Create: capacity: int -> 'T array
     [<Emit("$0.from($1)")>]
     abstract From: 'T seq -> 'T array
 
-let map (f: 'T->'U) (source: 'T[]) (cons: ArrayCons) =
+[<Emit("$0.push($1)")>]
+let private push (array: 'T[]) (item: 'T): int = jsNative
+
+[<Emit("$0.slice($1, $2)")>]
+let private slice (array: 'T[]) (``begin``: int) (``end``: int): 'T[] = jsNative
+
+[<Emit("$0.slice($1)")>]
+let private sliceFrom (array: 'T[]) (``begin``: int): 'T[] = jsNative
+
+[<Emit("$0.splice($1, $2)")>]
+let private splice (array: 'T[]) (start: int) (deleteCount: int): 'T[] = jsNative
+
+// TODO: Should we reuse declarations from Import.JS instead of raw emits?
+let inline private spliceN<'T> (array: 'T[]) (start: int) (deleteCount: int) =
+    (box array :?> JS.Array<_>).splice(float(start), float(deleteCount)) |> box :?> 'T[]
+
+[<Emit("$0.indexOf($1)")>]
+let private indexOf (array: 'T[]) (item: 'T): int = jsNative
+
+let inline private isTypedArray arr = JS.ArrayBuffer.isView arr :?> bool
+
+[<Emit("[].concat.apply([], $0)")>]
+let private dynamicArrayConcatImpl (arrays: 'T[][]): 'T[] = jsNative
+
+[<Emit("$0.set($1, $2)")>]
+let private typedArraySetImpl (target: obj) (source: obj) (offset: float) = jsNative
+
+[<Emit("$1.find($0)")>]
+let inline private findImpl predicate (array: 'T[]): 'T option = jsNative
+
+[<Emit("$1.findIndex($0)")>]
+let inline private findIndexImpl predicate (array: 'T[]): int = jsNative
+
+// TODO: Should we tend to use native js counterparts as much as possible
+// (and how do Replacements nativeArrayFunctions relate?), i.e. chase performance,
+// or should we do the opposite and implement all the function in F# (with small exceptions when typed arrays chime in),
+// go for correcntess for starters.
+[<Emit("$0.concat($1)")>]
+let append (array1: 'T[]) (array2: 'T[]): 'T[] = jsNative
```

Comment:
If the semantics are the same, I tend to think native JS methods should produce less code and yield better performance ([though this is not always true](https://twitter.com/FableCompiler/status/890488515926122496)). Yes, the Replacements module [use native JS functions for arrays when possible](https://github.com/fable-compiler/Fable/blob/a844ca0621548c3dfb20229f9c7c04a030856207/src/dotnet/Fable.Compiler/Replacements.fs#L2030-L2037).

Reply:
Oh, that feeling when you thought you knew that js<->native transitions do hurt performance, but underestimate consequences by a magnitude... I think it's better to leave things as is right now and consider optimizing anything only after benchmark results chime in. Do you agree?

Reply:
Sounds good! Yes, let's not worry too much about performance at this point, but leave a comment where in doubt so we can benchmark late :smile: